vllm/timings/timer.py

```diff
@@ -0,0 +1,81 @@
+import logging
+import time
+from contextlib import contextmanager
+from threading import RLock
+
+import numpy
+
+__all__ = ["Timer"]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class Timer:
+    """Timer to log timings during inference requests. Should be used through 
+        timings.utils.get_singleton_manager()
+    """
+    _instance = None
+
+    def __init__(self, avg_after_iterations: int, enable_logging: bool):
+        """
+        :param avg_after_iterations: the number of iterations that have to occur
+            before the time measurements are averaged and displayed to the user
+        :param enable_logging: whether or not time logging is enabled
+        """
+        self.enable_logging = enable_logging
+        self.measurements = {}
+        self.avg_after_iterations = avg_after_iterations
+        self.current_iter = {}
+        self.lock = RLock()
```

Comment:
Is there any way we could make the timer be "owned" by the `LLMEngine` / `_AsyncLLMEngine` ?

I don't actually think we need a `lock` since everything we are timing is inside the `_AsyncLLMEngine.step_async` function or `LLMEngine.step` function (which both run inside the `run_engine_loop`  (and thus there are no race conditions because only one thread is accessing the object), but having the Timer owned by the LLMEngine would make this explicit

Reply:
That would be different than having it owned by the `StatLogger`, correct? I think what you described above with the `StatLogger` makes more sense, in terms of what owns the timers/how the averages/measurements are used from the timer.

The idea with having it as a singleton which is first initialized by the entrypoint was to prevent it from being intrusive and also make it such that we can easily use it both within the server or outside the server without having to add a ton of additional code. I think having it as part of the `StatLogger` would still reach this goal.

Reply:
@robertgshaw2-neuralmagic Updated such that timer init and average computation is now handled by the `StatLogger`. The `StatLogger` is configured by the `--disable-log-stats` command already so I removed the cli arg added such for the timer. Dont think it makes sense to have both.